### PR TITLE
[PyTorch TB] Write raw tensor as tensor_proto

### PR DESCRIPTION
**Pull request by sethlee0111**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #104908
* #105027

This is the first diff to support logging of raw tensors for [TensorBoard Intermediate Logging](https://www.internalfb.com/intern/wiki/TensorBoard/Intermediate_Logging/)

Ultimately, we aim to support the feature where store full tensor is stored as a tensor protobuf to TB. Protobuf contains shape, dtype, and elements of the given tensor.

1. add `tensor_proto()` to `summary.py` which takes a tensor and convert to protobuf
2. add `add_tensor()` to `writer.py`
3. formatting changes introduced by `arc lint`
-------------

Differential Revision: [D47302017](https://our.internmc.facebook.com/intern/diff/D47302017/)

### CODE REVIEW

File: torch/utils/tensorboard/summary.py [hunk 1]
```diff
@@ -303,6 +325,75 @@ def scalar(name, tensor, collections=None, new_style=False, double_precision=Fal
...
+    if tensor.dtype in type_map:
+        proto_val_field = type_map[tensor.dtype][1]
+        tensor_proto_args = {
+            "dtype": type_map[tensor.dtype][0],
+            "dim": [
+                TensorShapeProto.Dim(size=tensor.shape[i]) for i in range(tensor.dim())
+            ],
+            proto_val_field: torch.view_as_real(tensor).tolist()
+            if proto_val_field == "scomplex_val" or proto_val_field == "dcomplex_val"
+            else tensor.reshape(-1).tolist(),
```
Comment (vadimkantorov) [inline torch/utils/tensorboard/summary.py:383]: here a bit more idiomatic and modern might be `.flatten()` instead of reshape(-1)

Comment (sethlee0111) [reply torch/utils/tensorboard/summary.py]: Updated, thanks!

Comment (sethlee0111) [reply torch/utils/tensorboard/summary.py]: Added checks for `numel() == 1` and `numel() == 0`

File: torch/utils/tensorboard/summary.py [hunk 2]
```diff
@@ -323,6 +326,85 @@ def scalar(name, tensor, collections=None, new_style=False, double_precision=Fal
...
+      ValueError: If tensor is too big to be converted to protobuf, or
+                     tensor data type is not supported
+    """
+    if tensor.numel() * tensor.itemsize >= (1 << 31):
+        raise ValueError(
+            "tensor is bigger than protocol buffer's hard limit of 2GB in size"
+        )
+
+    # type maps: torch.Tensor type -> (protobuf type, protobuf val field)
+    type_map = {
```
Comment (kunalb) [inline torch/utils/tensorboard/summary.py:348]: make this a private constant map outside the function? _TENSOR_TYPE_MAP = ... or something?

File: torch/utils/tensorboard/summary.py [hunk 3]
```diff
@@ -323,6 +326,85 @@ def scalar(name, tensor, collections=None, new_style=False, double_precision=Fal
...
+        torch.complex128: ("DT_COMPLEX128", "dcomplex_val"),
+        torch.cdouble: ("DT_COMPLEX128", "dcomplex_val"),
+        torch.uint8: ("DT_UINT8", "uint32_val"),
+        torch.quint8: ("DT_UINT8", "uint32_val"),
+        torch.quint4x2: ("DT_UINT8", "uint32_val"),
+    }
+
+    if tensor.dtype in type_map:
+        proto_val_field = type_map[tensor.dtype][1]
+        proto_val_contents = (
```
Comment (kunalb) [inline torch/utils/tensorboard/summary.py:379]: For readability, I'd still this make this explicit
proto_val_contents = [] 
if proto_val_field == ...:
    proto_val_contents = ...
elif tensor.numel() == 1
    proto_val_contents  = ...

and so on

**Review by kunalb — APPROVED** [review]
Generally looks good to me, thanks for adding tests! some nits inline

File: torch/utils/tensorboard/summary.py (hunk 2 above)
Comment (sethlee0111) [reply torch/utils/tensorboard/summary.py]: Makes sense, applied changes.

File: torch/utils/tensorboard/summary.py (hunk 3 above)
Comment (sethlee0111) [reply torch/utils/tensorboard/summary.py]: I see what you mean. Made changes.

**Pull request closed without merging**